hudi-utilities/src/main/java/org/apache/hudi/utilities/deltastreamer/DeltaSync.java

```diff
@@ -330,31 +330,28 @@ public void refreshTimeline() throws IOException {
-          resumeCheckpointStr = Option.of(cfg.checkpoint);
-        } else if (!StringUtils.isNullOrEmpty(commitMetadata.getMetadata(CHECKPOINT_KEY))) {
-          //if previous checkpoint is an empty string, skip resume use Option.empty()
-          resumeCheckpointStr = Option.of(commitMetadata.getMetadata(CHECKPOINT_KEY));
-        } else if (HoodieTimeline.compareTimestamps(HoodieTimeline.FULL_BOOTSTRAP_INSTANT_TS,
-            HoodieTimeline.LESSER_THAN, lastCommit.get().getTimestamp())) {
-          // if previous commit metadata did not have the checkpoint key, try traversing previous commits until we find one.
-          Option<String> prevCheckpoint = getPreviousCheckpoint(commitTimelineOpt.get());
-          if (prevCheckpoint.isPresent()) {
-            resumeCheckpointStr = prevCheckpoint;
-          } else {
+        // if previous commit metadata did not have the checkpoint key, try traversing previous commits until we find one.
+        Option<HoodieCommitMetadata> commitMetadataOption = getLatestCommitMetadataWithValidCheckpointInfo(commitTimelineOpt.get());
+        if (commitMetadataOption.isPresent()) {
+          HoodieCommitMetadata commitMetadata = commitMetadataOption.get();
+          if (cfg.checkpoint != null && (StringUtils.isNullOrEmpty(commitMetadata.getMetadata(CHECKPOINT_RESET_KEY))
+              || !cfg.checkpoint.equals(commitMetadata.getMetadata(CHECKPOINT_RESET_KEY)))) {
+            resumeCheckpointStr = Option.of(cfg.checkpoint);
+          } else if (!StringUtils.isNullOrEmpty(commitMetadata.getMetadata(CHECKPOINT_KEY))) {
+            //if previous checkpoint is an empty string, skip resume use Option.empty()
+            resumeCheckpointStr = Option.of(commitMetadata.getMetadata(CHECKPOINT_KEY));
+          } else if (HoodieTimeline.compareTimestamps(HoodieTimeline.FULL_BOOTSTRAP_INSTANT_TS,
+              HoodieTimeline.LESSER_THAN, lastCommit.get().getTimestamp())) {
             throw new HoodieDeltaStreamerException(
                 "Unable to find previous checkpoint. Please double check if this table "
                     + "was indeed built via delta streamer. Last Commit :" + lastCommit + ", Instants :"
                     + commitTimelineOpt.get().getInstants().collect(Collectors.toList()) + ", CommitMetadata="
                     + commitMetadata.toJsonString());
           }
-        }
-        // KAFKA_CHECKPOINT_TYPE will be honored only for first batch.
-        if (!StringUtils.isNullOrEmpty(commitMetadata.getMetadata(CHECKPOINT_RESET_KEY))) {
-          props.remove(KafkaOffsetGen.Config.KAFKA_CHECKPOINT_TYPE.key());
+          // KAFKA_CHECKPOINT_TYPE will be honored only for first batch.
+          if (!StringUtils.isNullOrEmpty(commitMetadata.getMetadata(CHECKPOINT_RESET_KEY))) {
+            props.remove(KafkaOffsetGen.Config.KAFKA_CHECKPOINT_TYPE.key());
+          }
         }
       }
```

Comment:
Can we extract this whole logic of getting resume checkpoint to a single method for better readability?